plugins/sentinelone/komand_sentinelone/tasks/monitor_activities_and_events/task.py

```diff
@@ -0,0 +1,327 @@
+            return [], None, total_forbidden_responses
+        response_json = extract_json(response)
+        logs = response_json.get("data", [])
+        next_cursor = response_json.get("pagination", {}).get("nextCursor")
+        return logs, next_cursor, total_forbidden_responses
+
+    def get_latest_timestamp(self, logs, latest_timestamp_string) -> Optional[str]:
+        """
+        Return the latest timestamp from a sorted list of logs. If no logs exist, use the last provided timestamp
+        """
+        # If no logs returned, return existing timestamp or None if never previously set
+        if len(logs) == 0:
+            if latest_timestamp_string == STARTING_TIMESTAMP:
+                return None
+            return latest_timestamp_string
+        new_timestamp_string = logs[-1].get("createdAt")
+        if new_timestamp_string is None:
+            new_timestamp_string = logs[-1].get("threatInfo", {}).get("createdAt")
+        new_timestamp = datetime.strptime(new_timestamp_string, DATE_TIME_FORMAT).astimezone(timezone.utc)
+        latest_timestamp = datetime.strptime(latest_timestamp_string, DATE_TIME_FORMAT).astimezone(timezone.utc)
+        if new_timestamp > latest_timestamp:
+            latest_timestamp_string = new_timestamp_string
+        return latest_timestamp_string
+
+    def determine_next_pagination_cycle(self, state: dict, lookback_timestamp: str, current_run_timestamp: str) -> bool:
+        """
+        Set the last run timestamp in state based upon pagination cycle.
+        Return true if pagination expected next cycle.
+        """
+        has_more_pages = any(
+            [state.get(ACTIVITIES_PAGE_CURSOR), state.get(EVENTS_PAGE_CURSOR), state.get(THREATS_PAGE_CURSOR)]
+        )
+        if has_more_pages:
+            self.logger.info(
+                "Pagination token was returned. Not updating last run timestamp to current time until query completed."
+            )
+            state[LAST_RUN_TIMESTAMP] = lookback_timestamp
+        else:
+            state[LAST_RUN_TIMESTAMP] = current_run_timestamp
+        return has_more_pages
+
+    def get_lookback_values(
+        self, is_initial_run: bool, custom_config: dict = {}, last_timestamp: str = None
+    ) -> Tuple[str, str]:
+        """
+        Determine cutoff hours from running state, or supplied custom config cutoff hours which take precedent.
+        Determine a lookback date from last run timestamp or, if not available, from the determined cutoff hours.
+        If a custom lookback date is supplied in the custom config, use this as a lookback date for initial runs.
+        Return the current date and the calculated lookback date time
+        """
+        # Get the current time
+        current_date_time = datetime.now(timezone.utc)
+        lookback_date = None
+        if custom_config:
+            self.logger.info("Custom config detected")
+        # Set the lookback date to the last timestamp if it exists
```

Comment:
Suggestion: For readability, can you add a blank line before each of the comments in this function?